Introduce alpha version of Back in Stock Notifications

xristos3490 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR introduces an alpha version of back-in-stock notifications. There is still remaining work to do when capacity permits us.

The feature is disabled by default using a constant gate named `WOOCOMMERCE_BIS_ALPHA_ENABLED`. To enable it, add the following line to your `wp-config.php` file:
```php
define( 'WOOCOMMERCE_BIS_ALPHA_ENABLED', true );
```

Individual PRs include:
- #57062
- #58044
- #58202
- #58356
- #59006
- #59093
- #59047
- #58388
- #59331
- #59594
- #59492
- #59595
- #59470
- #59622

### Disclaimer

This is an alpha version with missing features and is not recommended for use in production environments.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Testing the dbDelta
1. Use an existing store.
2. Delete the options `woocommerce_version` and `woocommerce_db_version` using your preferred database manager (Clear your object cache if you're using one!).
3. Check out this branch.
4. Ensure that the tables `wp_wc_stock_notifications` and `wp_wc_stock_notificationmeta` are not included in the installed tables.
5. Create a new WooCommerce store with this branch.
6. Verify again that the installed tables do not include the bis-related ones.

#### (Bonus) Testing that classes aren't loaded
1. Use xDebug to set a breakpoint on the `Automattic\WooCommerce\Internal\StockNotifications::__construct()` method.
2. Refresh and confirm this never runs in both the front-end and back-end.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

xristos3490 committed:
MIC: Create base infrastructure and DB tables for Back in Stock Notifications (#57062)

* Add base service and database schema

* Make linter happy

* Linter ignore

* Rename datastore class

* Add basic CRUD for notifications, metadata, and a helper for logs

* Make linter happy again

* Shorten table names, suffix dates with gmt, introduce dedicated meta store

* Fix docblock on reusable datastore

* Lint fixes

* Lint fixes

* Update service provider

* Lint fixes

* Lint fixes

* Lint fixes

* Lint fixes

* Fix dbDelta issues with capital letters and phpcs fixes

* Add specific query formats

* Remove object keyword refs

* Remove meta cache duplication in favor of WC_Data level caching

* Improvements in stock notifications CRUD and introduce WP_Error returns

* Docblock improvements

* Create the main controller and wire the data store

* Fix data stores for basic CRUD and add unit tests

* Add metadata handling and unit tests

* Lint fixes

* Organize unit tests and align create and modified dates

* Alias main controller init and make linter happy

* One more lint fix

* Imporvements and new db table indexes

* Clean up notification unit tests

* Add metadata unit tests

* Rename the event log helper

* Add a factory, unit tests, and move notifications tests

* Lint fixes

* Make linter happy

* Add a query method

* Create data store for logs

* Lint fixes

* Lint fixes and fix the init data store of logs

* Fix error in log params

* Use the main datastore for tests

* Use specified date modified value and some polishing

* Fix unit tests by applying changes on the object after saving the meta

* Lint fixes

* Setup timezone for tests to control output

* Avoid error logging in unit tests

* Fix timezones when reading data from DB

* Lint fixes

* Reorg the table names into getters in data stores

* Minor fixes on the scheme

* Move delete functions in each data store

* Lint fixes

* Fix unit tests

* Remove comment ignores

* Use assertCount

* Fix unit tests

* Add args and return types in activity log data store

* Added args and return types in notifications data store where applicable

* Fix the datastore docblock in activity logs tests

* Differentiate dates for better coverage

* Remove text based tests

* Check non-existance after deleting meta but not saved yet

* One more rename of note to payload

* Modify the database schema and fix the tests

* Make linter happy

* Add the id at the batch processing index

* Cleanup is_queued

* Add docblock details on notification statuses

* Add all param types for date-related setters

* Add tests for query types

* Improve performance on the query method and the model instantiation

* Make linter happy

* Fix the NotificationCancellationSource enum

* Add type int in factory

* Make NotificationStatus enums public

* Add types on Notification setters

* simplify db index statement

* Set the cancellation_source db size and use array_key_exists on update

* Protect from failed deletions

* Review on datastore

* Unit Tests FTW

* Improve deletion checkings

* Rename controller to StockNotifications

* Fix the main entrypoint

* Fix typo

xristos3490 committed:
MIC: Introduce Email Templates for Back In Stock Notifications (#58044)

* Add base service and database schema

* Make linter happy

* Linter ignore

* Rename datastore class

* Add basic CRUD for notifications, metadata, and a helper for logs

* Make linter happy again

* Shorten table names, suffix dates with gmt, introduce dedicated meta store

* Fix docblock on reusable datastore

* Lint fixes

* Lint fixes

* Update service provider

* Lint fixes

* Lint fixes

* Lint fixes

* Lint fixes

* Fix dbDelta issues with capital letters and phpcs fixes

* Add specific query formats

* Remove object keyword refs

* Remove meta cache duplication in favor of WC_Data level caching

* Improvements in stock notifications CRUD and introduce WP_Error returns

* Docblock improvements

* Create the main controller and wire the data store

* Fix data stores for basic CRUD and add unit tests

* Add metadata handling and unit tests

* Lint fixes

* Organize unit tests and align create and modified dates

* Alias main controller init and make linter happy

* One more lint fix

* Imporvements and new db table indexes

* Clean up notification unit tests

* Add metadata unit tests

* Rename the event log helper

* Add a factory, unit tests, and move notifications tests

* Lint fixes

* Make linter happy

* Add a query method

* Create data store for logs

* Lint fixes

* Lint fixes and fix the init data store of logs

* Fix error in log params

* Use the main datastore for tests

* Use specified date modified value and some polishing

* Fix unit tests by applying changes on the object after saving the meta

* Lint fixes

* Setup timezone for tests to control output

* Avoid error logging in unit tests

* Fix timezones when reading data from DB

* Lint fixes

* Reorg the table names into getters in data stores

* Minor fixes on the scheme

* Move delete functions in each data store

* Lint fixes

* Fix unit tests

* Remove comment ignores

* Use assertCount

* Fix unit tests

* Add args and return types in activity log data store

* Added args and return types in notifications data store where applicable

* Fix the datastore docblock in activity logs tests

* Differentiate dates for better coverage

* Remove text based tests

* Check non-existance after deleting meta but not saved yet

* One more rename of note to payload

* Introduce the email templates

* Simplify and reorg templates and controllers

* Lint fixes

* Add plain templates

* Handle plain mode in template functions

* Lint fixes

* Lint fixes

* Reorg of files, renamed the email classes, and introduced a single entrypoint on the Controller class

* Lint fixes

* Lint fixes

* revert unrelated changes

* Rename email titles

* Modify the database schema and fix the tests

* Make linter happy

* Add the id at the batch processing index

* Fix into->intro type

* Rename custom tax location method

* Fix verify template

* Improve getter for formatted variations and add a unit test

* Add unit tests for get_product_permalink/name

* Rename emails id and classes for better semantics

* Lint fixes on tests

* Lint fixes on tests

* Reorg namespaces into Emails, and move the templates into the email controller

* Rename EmailManager, move two emails in transactional emails actions and modify trigger points

* Lint

* Fix customer data hook

* Renamed confirm email to verified, and added better docblock for the customer data hook

* Simplify and repurpose get_product_formatted_variation_list; handle emails rendering on the spot

* Rename the template

* Lint

* Cleanup renamed files

* Revert minor change on the provider

jimjasson committed:
MIC: Add Customer Stock Notifications settings (#58202)

* Add base service and database schema

* Make linter happy

* Linter ignore

* Rename datastore class

* Add basic CRUD for notifications, metadata, and a helper for logs

* Make linter happy again

* Shorten table names, suffix dates with gmt, introduce dedicated meta store

* Fix docblock on reusable datastore

* Lint fixes

* Lint fixes

* Update service provider

* Lint fixes

* Lint fixes

* Lint fixes

* Lint fixes

* Fix dbDelta issues with capital letters and phpcs fixes

* Add specific query formats

* Remove object keyword refs

* Remove meta cache duplication in favor of WC_Data level caching

* Improvements in stock notifications CRUD and introduce WP_Error returns

* Docblock improvements

* Create the main controller and wire the data store

* Fix data stores for basic CRUD and add unit tests

* Add metadata handling and unit tests

* Lint fixes

* Organize unit tests and align create and modified dates

* Alias main controller init and make linter happy

* One more lint fix

* Imporvements and new db table indexes

* Clean up notification unit tests

* Add metadata unit tests

* Rename the event log helper

* Add a factory, unit tests, and move notifications tests

* Lint fixes

* Make linter happy

* Add a query method

* Create data store for logs

* Lint fixes

* Lint fixes and fix the init data store of logs

* Fix error in log params

* Use the main datastore for tests

* Use specified date modified value and some polishing

* Fix unit tests by applying changes on the object after saving the meta

* Lint fixes

* Setup timezone for tests to control output

* Avoid error logging in unit tests

* Fix timezones when reading data from DB

* Lint fixes

* Reorg the table names into getters in data stores

* Minor fixes on the scheme

* Move delete functions in each data store

* Lint fixes

* Fix unit tests

* Remove comment ignores

* Use assertCount

* Fix unit tests

* Add args and return types in activity log data store

* Added args and return types in notifications data store where applicable

* Fix the datastore docblock in activity logs tests

* Differentiate dates for better coverage

* Remove text based tests

* Check non-existance after deleting meta but not saved yet

* One more rename of note to payload

* Introduce the email templates

* Simplify and reorg templates and controllers

* Lint fixes

* Add plain templates

* Handle plain mode in template functions

* Lint fixes

* Lint fixes

* Reorg of files, renamed the email classes, and introduced a single entrypoint on the Controller class

* Lint fixes

* Lint fixes

* revert unrelated changes

* Rename email titles

* Modify the database schema and fix the tests

* Make linter happy

* Add the id at the batch processing index

* Fix into->intro type

* MIC: Introduce Customer Stock Notifications settings

* Lint fixes

* Improve unit tests

* Group Guest sign-up options

* Lint fixes

* Rename custom tax location method

* Fix verify template

* Remove unused variables

* Remove unused argument

* Lint fixes

* Improve getter for formatted variations and add a unit test

* Add unit tests for get_product_permalink/name

* Rename emails id and classes for better semantics

* Lint fixes on tests

* Lint fixes on tests

* Reorg namespaces into Emails, and move the templates into the email controller

* Rename EmailManager, move two emails in transactional emails actions and modify trigger points

* Lint

* Fix customer data hook

* - Fix conditional visibility of Guest Sign-up options.
- Remove Minimum Stock Quantity option.
- Add extra validations for warnings

* Removed type hints from filter callbacks

* Update option names and enable sign ups for new stores

* Update option names in unit test

* Lint

* Lint

* Review

* Lint review

* More lint fixes

* More lint fixes

* Share the controller

* Few more conflicts

* One last conflict

* On more

* Oops

* Update option names and unit tests

* Lint fixes

* Review

---------

Co-authored-by: Chris Lilitsas <1105590+xristos3490@users.noreply.github.com>

jimjasson committed:
MIC: Add Customer Stock Notifications menus and Notifications List Tables (#58356)

* Add base service and database schema

* Make linter happy

* Linter ignore

* Rename datastore class

* Add basic CRUD for notifications, metadata, and a helper for logs

* Make linter happy again

* Shorten table names, suffix dates with gmt, introduce dedicated meta store

* Fix docblock on reusable datastore

* Lint fixes

* Lint fixes

* Update service provider

* Lint fixes

* Lint fixes

* Lint fixes

* Lint fixes

* Fix dbDelta issues with capital letters and phpcs fixes

* Add specific query formats

* Remove object keyword refs

* Remove meta cache duplication in favor of WC_Data level caching

* Improvements in stock notifications CRUD and introduce WP_Error returns

* Docblock improvements

* Create the main controller and wire the data store

* Fix data stores for basic CRUD and add unit tests

* Add metadata handling and unit tests

* Lint fixes

* Organize unit tests and align create and modified dates

* Alias main controller init and make linter happy

* One more lint fix

* Imporvements and new db table indexes

* Clean up notification unit tests

* Add metadata unit tests

* Rename the event log helper

* Add a factory, unit tests, and move notifications tests

* Lint fixes

* Make linter happy

* Add a query method

* Create data store for logs

* Lint fixes

* Lint fixes and fix the init data store of logs

* Fix error in log params

* Use the main datastore for tests

* Use specified date modified value and some polishing

* Fix unit tests by applying changes on the object after saving the meta

* Lint fixes

* Setup timezone for tests to control output

* Avoid error logging in unit tests

* Fix timezones when reading data from DB

* Lint fixes

* Reorg the table names into getters in data stores

* Minor fixes on the scheme

* Move delete functions in each data store

* Lint fixes

* Fix unit tests

* Remove comment ignores

* Use assertCount

* Fix unit tests

* Add args and return types in activity log data store

* Added args and return types in notifications data store where applicable

* Fix the datastore docblock in activity logs tests

* Differentiate dates for better coverage

* Remove text based tests

* Check non-existance after deleting meta but not saved yet

* One more rename of note to payload

* Introduce the email templates

* Simplify and reorg templates and controllers

* Lint fixes

* Add plain templates

* Handle plain mode in template functions

* Lint fixes

* Lint fixes

* Reorg of files, renamed the email classes, and introduced a single entrypoint on the Controller class

* Lint fixes

* Lint fixes

* revert unrelated changes

* Rename email titles

* Modify the database schema and fix the tests

* Make linter happy

* Add the id at the batch processing index

* Fix into->intro type

* MIC: Introduce Customer Stock Notifications settings

* Lint fixes

* Improve unit tests

* Group Guest sign-up options

* Lint fixes

* Rename custom tax location method

* Fix verify template

* Remove unused variables

* Remove unused argument

* Lint fixes

* Improve getter for formatted variations and add a unit test

* Add unit tests for get_product_permalink/name

* Rename emails id and classes for better semantics

* Lint fixes on tests

* Lint fixes on tests

* Reorg namespaces into Emails, and move the templates into the email controller

* Rename EmailManager, move two emails in transactional emails actions and modify trigger points

* Lint

* Fix customer data hook

* - Fix conditional visibility of Guest Sign-up options.
- Remove Minimum Stock Quantity option.
- Add extra validations for warnings

* Removed type hints from filter callbacks

* Update option names and enable sign ups for new stores

* Update option names in unit test

* Lint

* Lint

* Review

* Lint review

* More lint fixes

* More lint fixes

* Share the controller

* Initial commit to add controllers and Notifications List table

* Fix lint errors

* Few more conflicts

* One last conflict

* On more

* Add functionality to the Notifications list table

* Oops

* Lint fixes

* More lint fixes

* Fix function args

* Fix dependencies

* Update option names and unit tests

* Lint fixes

* Lint fixes

* Review

* Support filtering by month

* Lint fixes

* Support multiple statuses in query args

* Fix lint errors

* More lint fixes

* Handle bulk actions and deleting individual notifications

* Fix lint errors

* More lint fixes

* More lint fixes

* More lint fixes

* Remove inactive views, use pending/cancelled/sent instead

* Add JS script for confirmation when deleting a notification

* Lint fixes

* More lint fixes

* Lint and styling fixes

* Share the new classes in the Service Provider

* Support Order by functionality

* Lint fixes

* Improve blank state screen to look like Orders empty state screen

* Update screen ids

* More page slug updates

* Remove unused JS param

* More page url updates

* Fix SQL queries

* Review admin manager class

* Fix issues related to missing dates from the DB

* Updated logic around fetching notification months

* Add more context to translatable strings

* Move action handling to NotificationPage

* Fix lint errors

* Fill in missing creation date when updating

* Use a contractor instead of init and action instead of a section

* Lint fixes

* Handle missing creation date when updating

* Refactor the way notices are displayed

* Lint fixes

* Lint fixes

* Review

* Update folder name

* Review

* Delete plugins/woocommerce/src/Internal/StockNotifications/Admin/Views/html-admin-notifications.php

Remove template with capital V for views

* Cleanup templates

* Fix initial blankState

---------

Co-authored-by: Chris Lilitsas <1105590+xristos3490@users.noreply.github.com>

xristos3490 committed:
MIC: Fix admin notices on stock notifications list table (#59006)

Fix list-table notices structure

roccotripaldi committed:
MIC: remove customer data from stock notification via privacy erasers (#59093)

Currently when a customer request data erasure, Woo will anonymize customer data on orders. After this PR, Woo will also anonymize customer data on stock notifications.

jimjasson committed:
MIC: Support manual notifications creation + edit (#59047)

* MIC: Support manual notifications creation

* Fix lint errors

* More lint fixes

* More lint fixes

* More lint fixes + added email validation

* Ignore to-do comment

* Maybe now I can keep my to-do comment?

* Support editing notifications

* Lint fixes

* More lint fixes

* More lint fixes

* Review

* Remove whitespace

* Introduce a new NotificationsPage::add_notice

* Fix lint errors

* Reorder action handlers for perf

* Add an alert when deleting permanently

* Handle gracefully non-existing users

* Minor typo fix

* review on list table and remove the waiting since column

* Support pre-filled values when errors occure in form and introduce specific validation during creation

* Lint fixes

* Minor fix on the add_notice helper

* Lint fixes

* Lint fixes

* Ensure notice gets cleaned up if invalid

* Rearrange form edit-handler

* Add more data model updates on single-edit actions and remove the verification email for now

* Rename views to Templates

* Avoid using static functions

* few more defensive parts on notices handler

* Few more defensive parts

* Revert list table change

* Remove unneeded helper on EmailManager

---------

Co-authored-by: Chris Lilitsas <1105590+xristos3490@users.noreply.github.com>

xristos3490 committed:
Merge branch 'trunk' into mic/back-in-stock-notifications

xristos3490 committed:
MIC: Introduce stock tracking hooks and AS scheduling for Back in Stock Notifications (#58388)

* Add base service and database schema

* Make linter happy

* Linter ignore

* Rename datastore class

* Add basic CRUD for notifications, metadata, and a helper for logs

* Make linter happy again

* Shorten table names, suffix dates with gmt, introduce dedicated meta store

* Fix docblock on reusable datastore

* Lint fixes

* Lint fixes

* Update service provider

* Lint fixes

* Lint fixes

* Lint fixes

* Lint fixes

* Fix dbDelta issues with capital letters and phpcs fixes

* Add specific query formats

* Remove object keyword refs

* Remove meta cache duplication in favor of WC_Data level caching

* Improvements in stock notifications CRUD and introduce WP_Error returns

* Docblock improvements

* Create the main controller and wire the data store

* Fix data stores for basic CRUD and add unit tests

* Add metadata handling and unit tests

* Lint fixes

* Organize unit tests and align create and modified dates

* Alias main controller init and make linter happy

* One more lint fix

* Imporvements and new db table indexes

* Clean up notification unit tests

* Add metadata unit tests

* Rename the event log helper

* Add a factory, unit tests, and move notifications tests

* Lint fixes

* Make linter happy

* Add a query method

* Create data store for logs

* Lint fixes

* Lint fixes and fix the init data store of logs

* Fix error in log params

* Use the main datastore for tests

* Use specified date modified value and some polishing

* Fix unit tests by applying changes on the object after saving the meta

* Lint fixes

* Setup timezone for tests to control output

* Avoid error logging in unit tests

* Fix timezones when reading data from DB

* Lint fixes

* Reorg the table names into getters in data stores

* Minor fixes on the scheme

* Move delete functions in each data store

* Lint fixes

* Fix unit tests

* Remove comment ignores

* Use assertCount

* Fix unit tests

* Add args and return types in activity log data store

* Added args and return types in notifications data store where applicable

* Fix the datastore docblock in activity logs tests

* Differentiate dates for better coverage

* Remove text based tests

* Check non-existance after deleting meta but not saved yet

* One more rename of note to payload

* Introduce the email templates

* Simplify and reorg templates and controllers

* Lint fixes

* Add plain templates

* Handle plain mode in template functions

* Lint fixes

* Lint fixes

* Reorg of files, renamed the email classes, and introduced a single entrypoint on the Controller class

* Lint fixes

* Lint fixes

* revert unrelated changes

* Rename email titles

* Modify the database schema and fix the tests

* Make linter happy

* Add the id at the batch processing index

* Fix into->intro type

* Rename custom tax location method

* Fix verify template

* Improve getter for formatted variations and add a unit test

* Add unit tests for get_product_permalink/name

* Rename emails id and classes for better semantics

* Lint fixes on tests

* Lint fixes on tests

* Reorg namespaces into Emails, and move the templates into the email controller

* Rename EmailManager, move two emails in transactional emails actions and modify trigger points

* Lint

* Fix customer data hook

* Renamed confirm email to verified, and added better docblock for the customer data hook

* Simplify and repurpose get_product_formatted_variation_list; handle emails rendering on the spot

* Rename the template

* Lint

* Initial commit

* Refactor the notificationsProcessor and sync controller

* Introduce the stock sync and async processor controllers and tests

* Phpcs fixes

* More lint fixes

* More lint fixes

* More lint fixes

* More lint fixes

* Repurpose StockManagementHelper and cleanups

* Lint

* Prevent query for variations if no children

* Rename the helper for better semantics

* Fix AS args

* Fix AS args in tests

* Implement the attempt limit in the query

* Add unit tests for the attempt limit on query

* Cleanup debugs

* Make the unit test for skipped a bit more specific

* Fix parsing new AS args

* Final changes for the new AS args structure

* Use spread op

* Simplify throttle retun

* Specify array elements

* Rename and add DISTICT in query instead of array_unique

* Add runtime cache in config and rename hooks

* Add the EligibilityService

* Fix stocksynccontroller based on the new service

* Fix tests for stocksync and minor fix in notificationtests

* Use the aggregated hook in the processor

* Refactor logic into smaller components

* Lint fixes

* More lint fixes

* minor rearrangement

* Fix unit tests of the async processor

* Cleanup

* Add new tables in WC_Install::get_tables()

* Lint fixes

* Remove deprecated service provider

* Add typed class props on processor

* Polish sync hook

* Cache queue instance into prop

* Use local product_id ref

* harmonize logs group

* Oops

* Harmonize the logs group

* Properly handle unanavailable products

* Add another test for trashed products

* Fix the notice URL

xristos3490 committed:
Merge branch 'trunk' into mic/back-in-stock-notifications

roccotripaldi committed:
Customer Stock Notifications: daily cleanup task (#59331)

Adds a new setting under woo products so an admin can set the number of days to keep stale notifications. Runs a daily task to delete stale notifications.

Co-authored-by: roccotrip <rocco.tripaldi@automattic.com>

xristos3490 committed:
Fix product link in the Back-in-Stock Notifications List Table (#59594)

Fix product link in the List Table

roccotripaldi committed:
[MiC] introduce cryptography tokens for handling unsubscribe for customer stock notifications (#59492)

* add new field to data model

* Add methods for getting email links

* move key to metadata part 1

* update option name for consistency

* add data methods

* back out of stale changeset

* add some doc blocks

* Default expiration time should be an hour

* simplify methods based on feedback from xristos3490

* linting

* Unit tests

* linting

* Default should be 1 hour

* fix linting for doc blocks

* fix broken link in contributing doc

* Runtime cache on Config and minor fixes

* Add a HashHelper util

* Implement the helper on the data model

---------

Co-authored-by: roccotrip <rocco.tripaldi@automattic.com>
Co-authored-by: Chris Lilitsas <1105590+xristos3490@users.noreply.github.com>

xristos3490 committed:
Merge branch 'trunk' into mic/back-in-stock-notifications

xristos3490 committed:
MIC: Ensure that managed variations appear in the List Table when filtered by a product variable ID (#59595)

* Introduce the eligibility service as a DI on the list table and some cleanups

* Fetch the list table via the container

xristos3490 committed:
Introduce Product Page Form for Back in Stock Notifications (#59470)

* Add the frontend template

* Add assets and scripts

* Add frontend script

* Extend jQuery and polish the script

* Lint

* Fix opt-in request data checking

* Imporve a11y on the frontend form

* Lint

* A few more a11y tweaks and fix the product-level setting

* Minor styling fix

* Fix nonce failures when global signups are off

* Decouple settings namespace from utils

roccotripaldi committed:
MIC: Process unsubscribe and verification email links (#59622)

* Add a controller for processing email actions

* process actions if key matches

* add a few unit tests

---------

Co-authored-by: roccotrip <rocco.tripaldi@automattic.com>

xristos3490 committed:
Merge branch 'trunk' into mic/back-in-stock-notifications

xristos3490 committed:
Add constant gate on the main controller

xristos3490 committed:
Add constant gate on the sql schema

xristos3490 committed:
Skip unit tests

xristos3490 committed:
Lint

xristos3490 committed:
Bump versions from 0.0.0 to 10.2.0

xristos3490 committed:
Changelog

xristos3490 committed:
Bump template versions and fix plain filenames

xristos3490 committed:
Bump the frontend template

roccotripaldi commented:
Thanks for the ping! I'll test this out tomorrow.

PanosSynetos reviewed (approved):
Hey @xristos3490 - I tested this, and as discussed we should merge what we currently have, behind a constant and make sure that without the constant, no Back In Stock notifications code runs.

I tested on an existing local site (without the constant)
- Deleted the options
- The database tables are not created
- With xdebug it doesn't pass through the notifications constructor

I tested on an existing local site (with the constant)
- - Deleted the options
- The database tables were created
- With xdebug it passes through the notifications constructor

I tested on a fresh jurassic ninja (without the constant)
- The database tables are not created


Anything else I'd need to test, let me know!

Thanks for this work, let's resume it when this project gets prioritized

roccotripaldi reviewed (approved):
DB delta tested well on my end.

I tried with an existing store locally where I ran this branch, deleted the options, and activated and deactivated Woo. There were not any stock notification tables created. 

I also did the same for a newly created local site, and had the same results.

In both cases i did not see any mentions of Customer Stock Notifications in the admin menus.

DB table creation kicked off as expected when i set the constant to `true`, and I see the expected menu items in wp-admin.

Great work! Merge away.

jimjasson reviewed (approved):
Hey @xristos3490, 

I tested on a JN site that already had WooCommerce 10.0.4 installed. 

I updated to the version in your branch and confirmed that no extra database tables were created. Then, I added `define( 'WOOCOMMERCE_BIS_ALPHA_ENABLED', true );` to my `wp-config.php` file, I refreshed the page and confirmed that:
- `wp_wc_stock_notifications` and;
- `wp_wc_stock_notificationmeta`

were created.

naman03malhotra reviewed (approved):
I've followed the given instrucitons for my local site:

- I checked out to this branch and I can see no BIS realted tables got created.
- Added the flag (`define( 'WOOCOMMERCE_BIS_ALPHA_ENABLED', true );`) to `wp-config.php` 
- I don't observe any new tables.
- I flush the cache, and I deleted version options, I start seeing the BIS related tables

<img width="364" height="92" alt="Screenshot 2025-07-28 at 1 54 09 PM" src="https://github.com/user-attachments/assets/5d8d133a-35a1-46a0-956e-519ff6a34bc8" />
<img width="299" height="62" alt="Screenshot 2025-07-28 at 1 59 33 PM" src="https://github.com/user-attachments/assets/663d16a0-07e6-4798-b566-f68ef3229c52" />

Also, I created a new store with this branch, I see the tables created.

Nice work!! 🎉 🚢

PanosSynetos commented:
Hey @naman03malhotra 

> Also, I created a new store with this branch, I see the tables created.

Was this site created with the constant set to true? If yes, then it's ok for the tables to be created. If no, then we have an issue, tables should not be created without the constant

nerrad commented on plugins/woocommerce/includes/class-woocommerce.php:
```diff
@@ -336,6 +337,9 @@ private function init_hooks() {
 		$container->get( ComingSoonRequestHandler::class );
 		$container->get( OrderCountCacheService::class );
 		$container->get( EmailImprovements::class );
+		if ( defined( 'WOOCOMMERCE_BIS_ALPHA_ENABLED' ) && WOOCOMMERCE_BIS_ALPHA_ENABLED ) {
```
Using the relevant methods on the `Automattic\Jetpack\Constants` class would be preferable here (makes it easier to set the flags for the constant values when unit testing).

nerrad commented on plugins/woocommerce/phpunit.xml:
```diff
@@ -16,6 +16,8 @@
 			<directory suffix=".php">./tests/php</directory>
 			<exclude>./tests/php/helpers</exclude>
 			<exclude>./tests/php/src/Blocks/BlockPatterns/patterns</exclude>
+			<exclude>./tests/php/src/Internal/StockNotifications</exclude>
+			<exclude>./tests/php/src/Internal/DataStores/StockNotifications</exclude>
```
Why are these being excluded?

xristos3490 replied:
Not a solid reason. I initially started adding logic to exclude tests in each `setUp()` method, but then I thought this would be cleaner since there isn't a test path with the constant enabled at the moment.

An alternative would be to keep them on and adjust the constant gate in `bootstrap.php`?

nerrad replied:
I'm coming from the mindset that we should have tests execute for this new code anyway. This could help surface negative side effects from changes in various parts of the codebase BIS integrates with earlier. So I imagine the actual test suites for BIS will need to have the constant set (which is another reason to use `Automattic\Jetpack\Constants`).

nerrad commented on plugins/woocommerce/src/Internal/StockNotifications/StockNotifications.php:
```diff
@@ -0,0 +1,78 @@
+		add_filter( 'woocommerce_data_stores', array( $this, 'register_data_stores' ) );
+
+		$container = wc_get_container();
+		$container->get( EmailManager::class );
+		$container->get( StockSyncController::class );
+		$container->get( NotificationsProcessor::class );
+		$container->get( PrivacyEraser::class );
+		$container->get( DataRetentionController::class );
+		$container->get( EmailActionController::class );
+
+		$container->get( ProductPageIntegration::class );
+		$container->get( FormHandlerService::class );
```
These need to be instantiated on every request? Even if the classes have further refinement on hooks they provide callbacks for, if those hooks are only fired on specific requests it seems like there might be some more request optimization here?

xristos3490 replied:
Thanks for this question!

I use this method while keeping the constructor simple and focused on registering hooks only. More complex operations are handled via lazy-loaded services (e.g., SignupService, EligibilityService). In my mind, this was a trade-off between performance optimization and architectural simplicity.

There is also some context checking (i.e., is_admin) to load admin-related classes when needed, and I planned to experiment with an is_frontend helper.

However, if I understand correctly, you're referring to a more refined approach to the Subscriber pattern. Instead of instantiating all these services upfront, we could use lightweight hook subscribers that register simple callbacks, and those callbacks fetch the actual handler from the container _only_ when the hook fires?

So rather than:
```php
$container->get(EmailManager::class);
```
We'd have something like:
```php
class EmailHookSubscriber {
    public function __construct($container) {
        add_action('woocommerce_email_classes', array( $this, 'emailClasses' ) );
    }
    
    public function emailClasses( $classes ) {
       return wc_get_container()->get(EmailManager::class)->emailClasses( $classes );
    }
}
```

It does mean more classes to maintain, though. What do you think would be the better approach for this? I could create a debt for this to tackle first thing; maybe also do some profiling to help us determine the trade-off here.

nerrad replied:
It could be premature optimization and might not even be needed, so I don't think it blocks merge. Would be good to follow-up when resuming work with an investigation on whether the overhead of instantiating all classes on every route no matter the context has measurable impact.

xristos3490 replied:
Sounds like a plan! Thanks for raising this! 🙌

nerrad reviewed (commented):
I haven't reviewed everything but just took a quick glance to see if anything obvious popped up and I've left comments on those things.

One other thing to watch very closely for is any places where WP localization functions are invoked too early. They should only be invoked on `init` or later (otherwise there will be a `Function _load_textdomain_just_in_time was called incorrectly ...` warning emitted)

xristos3490 commented on plugins/woocommerce/src/Internal/DataStores/StockNotifications/StockNotificationsDataStore.php:
```diff
@@ -0,0 +1,625 @@
+		$max_index_length = $this->database_util->get_max_index_length();
+
+		$sql = "
+CREATE TABLE $table_name (
+	id bigint(20) unsigned NOT NULL AUTO_INCREMENT,
+	product_id bigint(20) unsigned NOT NULL,
+	user_id bigint(20) unsigned NOT NULL,
+	user_email varchar(100) NOT NULL,
+	status varchar(20) NOT NULL DEFAULT 'pending',
+	date_created_gmt datetime NULL,
+	date_modified_gmt datetime NULL,
+	date_confirmed_gmt datetime NULL,
+	date_last_attempt_gmt datetime NULL,
+	date_notified_gmt datetime NULL,
+	date_cancelled_gmt datetime NULL,
+	cancellation_source varchar(30) NULL,
+	PRIMARY KEY  (id),
+	KEY product_status_attempt (product_id, status, date_last_attempt_gmt, id),
+	KEY user_lookup (user_id, product_id, status),
+	KEY email_lookup (user_email, product_id, status)
+) $collate;
```
User email is 100 characters long to match WordPress Core's wp_users table, which uses the same length.

xristos3490 committed:
Fix the options DOM in the ListTable

xristos3490 committed:
Avoid str_contains for php7.4 support

xristos3490 committed:
Merge branch 'trunk' into mic/back-in-stock-notifications

xristos3490 committed:
User Jetpack constants for feature gating

xristos3490 committed:
Bring back the tests

xristos3490 committed:
Remove empty lines

xristos3490 committed:
Use Jetpack Constants in all cases

xristos3490 commented:
Thank you all for your reviews. Special thanks to @jimjasson and @roccotripaldi for their contributions to this feature!

I have:

- Incorporated critical feedback from CodeRabbit.
- Re-activated all tests.
- Added Jetpack's Constants for feature flags.

This one should be ready for merge. Feel free to look again if you wish!

cc @PanosSynetos @nerrad @jamesckemp 

Also, fun fact: CodeRabbit put together a diagram from this PR. Not perfect, but still quite close!

```mermaid
sequenceDiagram
    participant Customer
    participant FrontendForm
    participant SignupService
    participant Notification
    participant Admin
    participant AdminUI
    participant NotificationDataStore
    participant StockSyncController
    participant JobManager
    participant NotificationsProcessor
    participant EmailManager
    participant WooCommerce

    Customer->>FrontendForm: Submit back-in-stock form
    FrontendForm->>SignupService: Validate & process signup
    SignupService->>Notification: Create notification (pending/active)
    Notification->>NotificationDataStore: Persist notification
    Admin->>AdminUI: Manage notifications (CRUD, filter)
    AdminUI->>NotificationDataStore: Query/update/delete notifications

    StockSyncController->>JobManager: Schedule async job on stock in
    JobManager->>NotificationsProcessor: Process notification batch
    NotificationsProcessor->>NotificationDataStore: Query eligible notifications
    NotificationsProcessor->>EmailManager: Send notification email(s)
    EmailManager->>Customer: Deliver notification email
    NotificationsProcessor->>NotificationDataStore: Update notification status
```

roccotripaldi reviewed (approved):
It's great to see my old friend Jetpack\Constants still in use.

I re-tested by:

- Creating a new store on site running this branch
- Deactivating Woo on an existing store, and then reactivating after replacing Woo with this branch
- In both cases, the db tables were not created, and there were no BIS related menu items in wp-admin
- I ran xdebug in the constructor of StockNotifications.php, and I can verify that the constructor is never called.

Send it!

xristos3490 committed:
Merge branch 'trunk' into mic/back-in-stock-notifications
